### PR TITLE
feat: ping

### LINKED ISSUE #251: Populate Fundamentals/Protocols doc for ping
## Done Criteria
- [ ] What is Ping
   - Request/response in p2p
   - Dialling/listening
- [ ] Ping in libp2p
   - Abstraction of the [official spec](https://github.com/libp2p/specs/blob/master/ping/ping.md)

## Why Important
- A better way to explicitly introduce the ping protocol.
- Provides a better reference point for other content.

## Notes
- Can use content distributed across the docs.
- Good term to add to the glossary.
- Follows https://github.com/libp2p/docs/issues/249

### PR DESCRIPTION
**Pull request by DannyS03**
## Context

- Adds content for libp2p ping
- closes #251

### Notes

Section this document lives under will become populated with corresponding PRs.

## Latest preview

Please view the latest Fleek preview [here](https://bafybeibnu7jbrxj5lsm3h4bnbf5vxyshhdgpzswshffc6qw3dvg7c4ss5q.on.fleek.co/concepts/introduction/protocols/ping/).

### CODE REVIEW

File: content/concepts/introduction/protocols/ping.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+---
+title: "Ping"
+description: "The ping protocol is a simple request response protocol."
+weight: 20
+---
+
+## What is Ping?
+
+Ping is a network utility used to test a node's reachability.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:9]: This is ICMP ping, not libp2p ping. You can only libp2p ping if you already have a libp2p connection. It’s worth noting here that these are completely different things though.

File: content/concepts/introduction/protocols/ping.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+The ping protocol is a simple request-response protocol that measures the RTT
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:10]: It’s not request-response, the stream is reused for future pings.

File: content/concepts/introduction/protocols/ping.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+for requests sent from an originating node to a destination node by echoing
+a request payload.
+
+## Ping in LibP2P
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:14]: Incorrect capitalization. It’s libp2p.

File: content/concepts/introduction/protocols/ping.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+In LibP2P, a peer opens a stream, sends a request with a payload of 32 random bytes,
+and the destination peer responds with 32 bytes on the same stream.
+
+An originating peer will only have one outbound stream at a time, and a destination peer
+accepts a maximum of two streams per peer. This is because the destination peer may view
+the originating peer as opening and closing the wrong stream as cross-stream behavior is
+non-linear.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:22]: I don’t understand this paragraph. I also don’t think that we need to talk about resource protection concerns in the documentation, that seems like something that should be discussed in the specs.

File: content/concepts/introduction/protocols/ping.md (hunk 2 above)
Comment (DannyS03) [reply content/concepts/introduction/protocols/ping.md]: The [spec](https://github.com/libp2p/specs/blob/master/ping/ping.md#protocol) mentions that is is request-response. I believe it is better to introduce the protocol as a liveness check, or state non-ICMP.

File: content/concepts/introduction/protocols/ping.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+---
+title: "Ping"
+description: "The ping protocol is a simple request response protocol."
+weight: 20
+---
+
+## What is Ping?
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:7]: I don't think we need 3 paragraphs about ICMP ping. I'd prefer to have a single sentence in the description of libp2p ping:
> The libp2p ping protocol is different from the ping command line utility (ICMP ping), as it requires an already established libp2p connection.

Comment (DannyS03) [reply content/concepts/introduction/protocols/ping.md:7]: As an introduction to ping it seems useful to have the additional context.

Comment (marten-seemann) [reply content/concepts/introduction/protocols/ping.md:7]: Yeah but this is confusion. We can't have an article about libp2p ping start with 3 paragraphs about something completely different that happens to have the same name.

Comment (DannyS03) [reply content/concepts/introduction/protocols/ping.md:7]: Updated

File: content/concepts/introduction/protocols/ping.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+Ping is a network utility that can be used as a simple liveness check
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:9]:
```suggestion
ICMP Ping is a network utility that can be used as a simple liveness check
```

Move this paragraph below the 1st paragraph in "Ping in libp2p"?

File: content/concepts/introduction/protocols/ping.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+for peers to quickly see if another peer is online and measure RTT.
+
+## Ping in libp2p
+
+The libp2p ping protocol is different from the ping command line utility
+([ICMP ping](https://en.wikipedia.org/wiki/Internet_Control_Message_Protocol)),
+as it requires an already established libp2p connection.
+
+A peer opens a stream, sends a request with a payload of 32 random
+bytes, and the destination peer responds with 32 bytes on the same stream.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:19]:
```suggestion
bytes, and the receiver responds with 32 bytes on the same stream.
```

File: content/concepts/introduction/protocols/ping.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+Peers can reuse a strean for future pings.
+
+Typically, a ping is sent over a stream after the initial protocol negotiation.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:22]: I don't think that's the case.

File: content/concepts/introduction/protocols/ping.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+---
+
+## Ping in libp2p
+
+ICMP Ping is a network utility that can be used as a simple liveness check
+for peers to quickly see if another peer is online and measure RTT.
+
+The libp2p ping protocol is different from the ping command line utility
+([ICMP ping](https://en.wikipedia.org/wiki/Internet_Control_Message_Protocol)),
+as it requires an already established libp2p connection.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:14]: I'd start with the libp2p ping protocol, and then contrast it to ICMP ping.

File: content/concepts/introduction/protocols/ping.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+A peer opens a stream, sends a request with a payload of 32 random
+bytes, and the receiver responds with 32 bytes on the same stream.
+Peers can reuse a strean for future pings.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:18]:
```suggestion
Peers can reuse the stream for future pings.
```

File: content/concepts/introduction/protocols/ping.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+The ping protocol ID is `/ipfs/ping/1.0.0`.
+
+### Example
+
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:23]: Can you call out here that this is what you get when you use kubo?
As a library, libp2p itself doesn't have a command line interface.

File: content/concepts/introduction/protocols/ping.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+> ICMP Ping is a network utility that uses ICMP packets to
+> check the connectivity and latency between two networked devices.
+> It is typically used to check the reachability of a host on an IP network and
+> to measure the round-trip time for messages sent from the originating host to a
+> destination host.
+
+A peer opens the stream, sends a request with a payload of 32 random
+bytes, and the receiver responds with 32 bytes on the same stream.
+Peers can reuse a strean for future pings.
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:23]:
```suggestion
A peer opens a new stream on an existing libp2p connection, and sends a ping request with a random 32 byte payload. The receiver echoes these 32 bytes back on the same stream. By measuring the time between
request and response the initiator can calculate the round-trip time of the underlying libp2p connection.
The stream can be reused for future pings from the initiator.
```

File: content/concepts/introduction/protocols/ping.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+The ping protocol ID is `/ipfs/ping/1.0.0`.
+
+### Example
+
+The following example is done using [Kubo IPFS](https://github.com/ipfs/kubo).
```
Comment (marten-seemann) [inline content/concepts/introduction/protocols/ping.md:29]:
```suggestion
[Kubo](https://github.com/ipfs/kubo) exposes a command line interface to ping other peers, which uses the libp2p ping protocol.
```

**Review by marten-seemann — APPROVED** [review]
lgtm modulo the 2 suggestions

**Pull request merged**